### PR TITLE
8315735: VerifyError when switch statement used with synchronized block

### PR DESCRIPTION
**Pull request by lahodaj**
Having code like:
```
    public static void main(String... args) {
        int i1 = 0 + switch (args.length) {
            default -> {
                synchronized (args) {
                    yield 1;
                }
            }
        };
    }
```

fails with a verification error at runtime. In the classfile, the synchronized block is basically a try-finally (which has an implicit "catch (any)" block), and so it needs the same handling as ordinary try statements.

(When there's a try statement inside a switch expression, the pre-existing stack is stored in local variables, and restored at return from the case. This is because catch clauses start with an empty stack.)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315735](https://bugs.openjdk.org/browse/JDK-8315735): VerifyError when switch statement used with synchronized block (**Bug** - P4)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to [26e29fe0](https://git.openjdk.org/jdk/pull/15584/files/26e29fe0c06732b4d1cffc118d820bec6a3b93b2)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15584/head:pull/15584` \
`$ git checkout pull/15584`

Update a local copy of the PR: \
`$ git checkout pull/15584` \
`$ git pull https://git.openjdk.org/jdk.git pull/15584/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15584`

View PR using the GUI difftool: \
`$ git pr show -t 15584`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15584.diff">https://git.openjdk.org/jdk/pull/15584.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15584#issuecomment-1707913221)

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java [hunk 1]
```diff
@@ -1285,6 +1285,11 @@ public void visitTry(JCTry tree) {
                     hasTry[0] = true;
                 }
 
+                @Override
+                public void visitSynchronized(JCSynchronized tree) {
+                    hasTry[0] = true;
+                }
+
```
Comment (forax) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java:1292]: You can make this code slightly more efficient by using the fact that 'var' is enable to reference the type of an anonymous class.

```java
private boolean hasTry(JCSwitchExpression tree) {
            var treeScanner = new TreeScanner() {
                private boolean hasTry;                
 
                @Override
                public void visitTry(JCTry tree) {
                    hasTry = true;
                }

                @Override
                public void visitSynchronized(JCSynchronized tree) {
                    hasTry = true;
                }

                @Override
                public void visitClassDef(JCClassDecl tree) {
                }
                @Override
                public void visitLambda(JCLambda tree) {
                }
            };
            treeScanner.scan(tree);
            return treeScanner.hasTry;
        }
```

**Comment by forax** [comment]
> (When there's a try statement inside a switch expression, the pre-existing stack is stored in local variables, and restored at return from the case. This is because catch clauses start with an empty stack.)

Technically, the stack is not empty, you have the exception on top of the stack.

**Comment by lahodaj** [comment]
> > (When there's a try statement inside a switch expression, the pre-existing stack is stored in local variables, and restored at return from the case. This is because catch clauses start with an empty stack.)
> 
> Technically, the stack is not empty, you have the exception on top of the stack.

Right, I meant items that are on stack at the time the exception occurs are cleared.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java (hunk 1 above)
Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java:1292]: nice suggestion

**Review by vicente-romero-oracle — APPROVED** [review]
looks good, please consider Remi's seems like a nice refactoring

File: src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java [hunk 2]
```diff
@@ -1278,16 +1278,17 @@ private void doHandleSwitchExpression(JCSwitchExpression tree) {
     }
     //where:
         private boolean hasTry(JCSwitchExpression tree) {
-            boolean[] hasTry = new boolean[1];
-            new TreeScanner() {
+            var scan = new TreeScanner() {
```
Comment (ExE-Boss) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java:1281]: I think `scanner` might be a better name for this:
```suggestion
            var scanner = new TreeScanner() {
```

Comment (forax) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java:1281]: I agree, the first guideline of the excellent guide written by Stuart Marks about the usage of 'var' is "Choose variable names that provide useful information [1].

[1] https://openjdk.org/projects/amber/guides/lvti-style-guide#G1

**Review by vicente-romero-oracle — APPROVED** [review]
looks good

File: src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java (hunk 2 above)
Comment (lahodaj) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java]: To be honest, I don't see a big difference between "scan", "scanner" and "treeScanner" - these describe the type of the variable, but not its purpose. So, I've used `hasTryScanner`, which conveys the purpose as well. Not a big reason to use less readable tricks with anonymous and var either - javac normally uses local classes to perform tasks like this.

Comment (forax) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java]: I agree, that we should not spend too much time on that code.
At the same time, `scan.scan(...)` is not very readable, I believe that `scanner` or `treeScanner` are better names.

**Pull request closed without merging**